lib/dao.js

```diff
@@ -2407,6 +2583,182 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+          'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = this.getConnector();
+  assert(typeof connector.replaceById === 'function',
+          'replaceById() must be implemented by the connector');
+
+  if (isPKMissing(Model, cb))
+    return cb.promise;
+
+  var model = Model.modelName;
+  var hookState = {};
+
+  // Convert the data to be plain object so that update won't be confused
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
+
+  if (id !== data[pkName]) {
+    var err = new Error('id property (' + pkName + ') ' +
+            'cannot be updated from ' + inst[pkName] + ' to ' + data[pkName]);
+    err.statusCode = 400;
+    process.nextTick(function() { cb(err); });
+    return cb.promise;
+  }
+
+  var context = {
+    Model: Model,
+    instance: inst,
+    hookState: hookState,
+    options: options
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    if (strict) {
+      applyStrictCheck(Model, strict, data, inst, validateAndCallConnector);
+    } else {
+      validateAndCallConnector(null, data);
+    }
+
+    function validateAndCallConnector(err, data) {
+      if (err) return cb(err);
+      data = removeUndefined(data);
+      // update instance's properties    
+      inst.setAttributes(data);
+
+      var doValidate = true;
+      if (options.validate === undefined) {
+        if (Model.settings.automaticValidation !== undefined) {
+          doValidate = Model.settings.automaticValidation;
+        }
+      } else {
+        doValidate = options.validate;
+      }
+
+      if (doValidate){
+        inst.isValid(function (valid) {
+          if (!valid) return cb(new ValidationError(inst), inst);
+
+          callConnector();
+        }, data);
+      } else {
+          callConnector();
+      }
+
+      function callConnector() {
+        var idNames = Model.definition.idNames();
+        var propKeys = Object.keys(Model.definition.properties);
+        var nonIdsPropKeys = propKeys.filter(function(i) {return idNames.indexOf(i) < 0;});
+        for (var i = 0; i < nonIdsPropKeys.length; i++) {
+          var p = nonIdsPropKeys[i];
+          inst[p] = null;
+        }
+        copyData(data, inst);
+        var typedData = convertDataPropertiesByType(data);
+        context.data = typedData;
+        
+        function replaceCallback(err, data) {
+          if (err) return cb(err);
+
+          var ctx = {
+            Model: Model,
+            instance: context.instance,
+            hookState: hookState,
+            data: context.data,
+            isNewInstance:false,
+            options: options
+          };
```

Comment:
You are correct, we don't provide `ctx.instance` for `loaded` hook.

From what I can tell by looking at other DAO methods, the loaded hooks has this context properties:

```js
          Model: ...,
          data: ...,
          isNewInstance: ...,
          hookState: ...,
          options: ...
        };
```

That's why I want to add the new methods to `support/describe-operation-hooks.js` and then check the output table to verify that new context objects are consistent with the existing ones.